File: src/main/java/com/hazza/LeetCodeByJava/SegmentOne/No98ValidateBinarySearchTree.java
package com.hazza.LeetCodeByJava.SegmentOne;

import com.hazza.LeetCodeByJava.datastructure.TreeNode;

import java.util.Stack;

/**
 * Created with longelliJ IDEA.
 * Description: https://leetcode.com/problems/validate-binary-search-tree/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-10-16
 * Time: 9:12 AM
 */
public class No98ValidateBinarySearchTree {
    // solution 1
    public boolean isValidBST(TreeNode root) {
        if (root == null) return true;
        return dfs(root, Long.MAX_VALUE, Long.MIN_VALUE);
    }

    private boolean dfs(TreeNode node, long maxVal, long minVal) {
        if (node == null) return true;
        if (node.val >= maxVal || node.val <= minVal) return false;
        return dfs(node.left, node.val, minVal) && dfs(node.right, maxVal, node.val);
    }

    // solution 2
    public boolean isValidBST1(TreeNode root) {
        if (root == null) return true;
        Stack<TreeNode> stack = new Stack<>();
        TreeNode pre = null, node = root;
        while (node != null || !stack.isEmpty()) {
            while (node != null) {
                stack.add(node);
                node = node.left;
            }
            node = stack.pop();
            if (pre != null && node.val <= pre.val) return false;
            pre = node;
            node = node.right;
        }

        return true;
    }
}

File: src/main/python/leetcode_by_python/tree/No98_Validate_Binary_Search_Tree.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-09-04
# https://leetcode.com/problems/validate-binary-search-tree/
import sys

from src.main.python.leetcode_by_python.data_structure.TreeNode import TreeNode


class Solution:
    def isValidBST1(self, root: TreeNode) -> bool:
        def dfs(node, minVal, maxVal):
            if not node:
                return True
            if node.val >= maxVal or node.val <= minVal:
                return False
            return dfs(node.left, minVal, node.val) and dfs(node.right, node.val, maxVal)

        return dfs(root, -sys.maxsize - 1, sys.maxsize)

    def isValidBST2(self, root: TreeNode) -> bool:
        if not root:
            return True
        stack = []

        pre, cur = None, root
        while cur or stack:
            while cur:
                stack.append(cur)
                cur = cur.left
            cur = stack.pop()
            if pre and pre.val >= cur.val:
                return False
            pre = cur
            cur = cur.right

        return True
